pkg/querysharding/util.go

```diff
@@ -0,0 +1,76 @@
+package querysharding
+
+import (
+	"encoding/base64"
+	"sync"
+
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/promql/parser"
+	"github.com/thanos-io/thanos/pkg/store/storepb"
+)
+
+const (
+	CortexShardByLabel = "__CORTEX_SHARD_BY__"
+)
+
+var (
+	buffers = sync.Pool{New: func() interface{} {
+		b := make([]byte, 0, 100)
+		return &b
+	}}
+)
+
+func InjectShardingInfo(query string, shardInfo *storepb.ShardInfo) (string, error) {
+	expr, err := parser.ParseExpr(query)
+	if err != nil {
+		return "", err
+	}
+	b, err := shardInfo.Marshal()
+	if err != nil {
+		return "", err
+	}
+	parser.Inspect(expr, func(n parser.Node, _ []parser.Node) error {
+		if selector, ok := n.(*parser.VectorSelector); ok {
+			selector.LabelMatchers = append(selector.LabelMatchers, &labels.Matcher{
+				Type:  labels.MatchEqual,
+				Name:  CortexShardByLabel,
+				Value: base64.StdEncoding.EncodeToString(b),
```

Comment:
Let's initialize `base64.StdEncoding.EncodeToString(b)` as a variable to not encode it multiple times when we have more than 1 `VectorSelector`.

Reply:
yup